lang/cs/Org.Apache.REEF.Network/NetworkService/Codec/StreamingCodecFunctionCache.cs

```diff
@@ -0,0 +1,203 @@
+            _writeAsyncFuncCache = new Dictionary<Type, Func<T, IDataWriter, CancellationToken, Task>>();
+            _streamingCodecType = typeof (IStreamingCodec<>);
+        }
+
+        /// <summary>
+        /// Creates the read delegate function of StreamingCodec from the message type
+        /// </summary>
+        /// <param name="messageType">Type of message</param>
+        /// <returns>The read delegate function</returns>
+        internal Func<IDataReader, T> ReadFunction(Type messageType)
+        {
+            Func<IDataReader, T> readFunc;
+
+            if (!_readFuncCache.TryGetValue(messageType, out readFunc))
+            {
+                AddCodecFunctions(messageType);
+                readFunc = _readFuncCache[messageType];
+            }
+
+            return readFunc;
+        }
+
+        /// <summary>
+        /// Creates the read async delegate function of StreamingCodec from the message type
+        /// </summary>
+        /// <param name="messageType">Type of message</param>
+        /// <returns>The read async delegate function</returns>
+        internal Func<IDataReader, CancellationToken, T> ReadAsyncFunction(Type messageType)
+        {
+            Func<IDataReader, CancellationToken, T> readFunc;
+
+            if (!_readAsyncFuncCache.TryGetValue(messageType, out readFunc))
+            {
+                AddCodecFunctions(messageType);
+                readFunc = _readAsyncFuncCache[messageType];
+            }
+
+            return readFunc;
+        }
+
+        /// <summary>
+        /// Creates the write delegate function of StreamingCodec from the message type
+        /// </summary>
+        /// <param name="messageType">Type of message</param>
+        /// <returns>The write delegate function</returns>
+        internal Action<T, IDataWriter> WriteFunction(Type messageType)
+        {
+            Action<T, IDataWriter> writeFunc;
+
+            if (!_writeFuncCache.TryGetValue(messageType, out writeFunc))
+            {
+                AddCodecFunctions(messageType);
+                writeFunc = _writeFuncCache[messageType];
+            }
+
+            return writeFunc;
+        }
+
+        /// <summary>
+        /// Creates the write async delegate function of StreamingCodec from the message type
+        /// </summary>
+        /// <param name="messageType">Type of message</param>
+        /// <returns>The write async delegate function</returns>
+        internal Func<T, IDataWriter, CancellationToken, Task> WriteAsyncFunction(Type messageType)
+        {
+            Func<T, IDataWriter, CancellationToken, Task> writeFunc;
+
+            if (!_writeAsyncFuncCache.TryGetValue(messageType, out writeFunc))
+            {
+                AddCodecFunctions(messageType);
+                writeFunc = _writeAsyncFuncCache[messageType];
+            }
+
+            return writeFunc;
+        }
+
+        private void AddCodecFunctions(Type messageType)
```

Comment:
So I definitely need cache so as not to use reflection to call the already existing codec. So here I chose to use Lambda expressions and delegates to solve the issue. Do you or @markusweimer  know any other easier way? If yes, I am ok to implement that.